src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -21,15 +25,52 @@ namespace Raven.Server.Documents.Sharding.Handlers
     public class ShardedIncomingReplicationHandler : AbstractIncomingReplicationHandler<TransactionContextPool, TransactionOperationContext>
     {
         private readonly ShardedDatabaseContext.ShardedReplicationContext _parent;
-        private readonly ReplicationQueue _replicationQueue;
+        private readonly ShardedOutgoingReplicationHandler[] _handlers;
+
+        public long LastEtag { get; set; }
+        public string LastReplicateChangeVector { get; set; }
 
         public ShardedIncomingReplicationHandler(TcpConnectionOptions tcpConnectionOptions, ShardedDatabaseContext.ShardedReplicationContext parent,
-            JsonOperationContext.MemoryBuffer buffer, ReplicationLatestEtagRequest replicatedLastEtag, ReplicationQueue replicationQueue)
+            JsonOperationContext.MemoryBuffer buffer, ReplicationLatestEtagRequest replicatedLastEtag)
             : base(tcpConnectionOptions, buffer, parent.Server, parent.DatabaseName, replicatedLastEtag, parent.Context.DatabaseShutdown, parent.Server.ContextPool)
         {
             _parent = parent;
-            _replicationQueue = replicationQueue;
             _attachmentStreamsTempFile = new StreamsTempFile("ShardedReplication" + Guid.NewGuid(), false);
+            var connectionInfo = IncomingConnectionInfo.FromGetLatestEtag(replicatedLastEtag);
+
+            _handlers = new ShardedOutgoingReplicationHandler[_parent.Context.ShardCount];
+            for (int i = 0; i < _handlers.Length; i++)
+            {
+                var node = new ShardReplicationNode { Shard = i, Database = ShardHelper.ToShardName(_parent.DatabaseName, i) };
+                var info = GetConnectionInfo(node);
+                if (info == null)
```

Comment:
can we get `null` here? what does it mean?

Reply:
Done